CI: silent merge check

m3dwards opened the pull request:
In this design there is a new GHA job merged into the main repo that adds check runs to each PR on whether they have a silent merge conflict or not.

The idea is that we have a job that pretty much runs 24/7 triggered every 6 hours or so and set to run for 5.5 hours, it could be set to run on a GHA runner rather than a Cirrus runner. Obviously running on a Cirrus runner (or even multiple runners) would dramatically speed up how many PRs can be checked for silent merge conflicts.

The job uses the Github checks API to add a passing or failing test run to each PR directly. The silent merge check job itself should always pass, it writes pass or fail checks to the PRs.

It will look through all open PRs and discard anything that has a git merge failure or a failing test with the thinking that there is no point checking something that can't be merged anyway. Then it will select PRs that have not had a silent merge check and one by one pull the merge branch of the PR and the full CI job "Previous releases" on that branch.

When all PRs have been checked it will then re-check a PR that was checked the longest time ago.

Sammie05 commented:
Nice work on this! The label-based trigger for PeriodicMergeCICheck is a clean approach.

One thought  though since this is driven by label events, do we need to consider cases where multiple labels are applied at once? Would the condition still behave as expected?

Also agree with the suggestion from @maflcko  having a vector of task names could as well improve flexibility

m3dwards commented:
> Other stuff to check would be that this works at all and doesn't just re-run an ancient commit, like [#32989 (comment)](https://github.com/bitcoin/bitcoin/pull/32989#issuecomment-3133536724)

Just ran a test to specifically test for this by updating the default (master) branch and adding and removing the label and it correctly picked an updated merge commit based on the PR merged into the updated master.

0xB10C reviewed (commented):
Have you considered doing something like the following in `.github/workflows/ci.yml` to avoid having to maintain separate files?

```yaml
name: CI

on:
  push:
  pull_request:
    types: [opened, reopened, synchronize, labeled]

jobs:
  conditional-job:
    if: >
      github.event_name == 'push' ||
      (
        github.event_name == 'pull_request' &&
        (
          github.event.action == 'opened' ||
          github.event.action == 'reopened' ||
          github.event.action == 'synchronize' ||
          (
            github.event.action == 'labeled' &&
            contains(github.event.label.name, 'PeriodicMergeCICheck')
          )
        )
      )
    runs-on: ubuntu-latest
    steps:
      - run: echo "Running CI job"
```

m3dwards commented:
> Have you considered doing something like the following in `.github/workflows/ci.yml` to avoid having to maintain separate files?

Could be a good shout, seems a bit of a verbose check if that has to be included in each job. I'll have a think about it. This is exactly the type of feedback that's good to consider.

m3dwards commented on .github/workflows/silentmerge.yml:
```diff
@@ -0,0 +1,99 @@
+            echo "::notice title=Runner Selection::Using GitHub-hosted runners"
+          fi
+
+  previous-releases:
+    name: 'Previous releases, depends DEBUG'
+    if: contains(github.event.label.name, 'PeriodicMergeCICheck')
```
It's a fair comment. I'll see what the ci-failed notification could look like but you could be right that this approach is less than ideal. @0xB10C has suggested perhaps merging this into the main CI file but might come with a verbose predicate before each job.

m3dwards committed:
ci: add silent merge check

Adds a periodic CI workflow that runs every 6 hours to detect silent
merge failures across all open PRs. For each PR, it fetches the
pre-built merge commit (pull/<n>/merge), runs a build and the tests,
and posts the result as a check run against the PR's head SHA. PRs
that have never been checked are prioritised (oldest first),followed
by those with the oldest existing check result.

maflcko commented on .github/workflows/silent-merge-check.yml:
```diff
@@ -0,0 +1,72 @@
+            echo "provider=gha" >> "$GITHUB_OUTPUT"
+            echo "::notice title=Runner Selection::Using GitHub-hosted runners"
+          fi
+  verify-prs:
```
nit: Forgot to set the timeout to 6h for this task?

maflcko reviewed (approved):
lgtm, thx!

maflcko commented on .github/workflows/silent-merge-check.yml:
```diff
@@ -0,0 +1,72 @@
+
+on:
+  schedule:
+    - cron: "0 0 * * 0"  # Sunday 00:00 UTC
```
should this not be `    - cron: "0 */6 * * *"`?

m3dwards replied:
Most likely, definitely the intention was to run every 6 hours for 5.5 hours.

maflcko replied:
Actually, an odd minute may be better to avoid the peak time around minute 0:

`    - cron: "37 */6 * * *"`

m3dwards replied:
https://www.upworthy.com/pick-a-random-number-between-100-you-probably-chose-37-and-there-s-a-big-reason-for-that/

maflcko commented on .github/silent-merge-check.py:
```diff
@@ -0,0 +1,215 @@
+    repo_root = Path(__file__).resolve().parent.parent
+    os.chdir(repo_root)
+    start_time = time.monotonic()
+    max_runtime_seconds = int(timedelta(hours=5, minutes=30).total_seconds())
```
nit: I think this could even be 3 hours (with a timeout and cron of 6 hours), to get a pull re-run every week, which should be enough?

maflcko commented:
are you still working on this?

m3dwards commented:
> are you still working on this?

yep, was waiting until the new CI provider was picked and merged but I guess there isn't any dependency there.

maflcko commented:
Yeah, this `runs-on: ubuntu-latest` (GHA), so there shouldn't be any blocker

m3dwards commented:
More concerned with the cache actions changing but can deal with that if and when it happens.

willcl-ark commented on .github/workflows/silent-merge-check.yml:
```diff
@@ -0,0 +1,74 @@
+    needs: [runners]
+    runs-on: ubuntu-latest
+    permissions:
+      checks: write
```
I think running the entire job with this permission would let a PR do some annoying things on GH as they have an authenticated write token to play with in run_all.sh?

Probably should just have this token in a seperate step or job where it's needed

m3dwards replied:
I'll have a think about this as you are right, the PR authors code will have access to the checks write token.

willcl-ark commented on .github/silent-merge-check.py:
```diff
@@ -0,0 +1,166 @@
+from datetime import datetime, timedelta
+from pathlib import Path
+
+MAX_RUNTIME = timedelta()
```
Is this intentional? surely we'd want here

```
timedelta(hours=5, minutes=30)
```

or read from an env var (from the main workflow)

m3dwards replied:
Yup, refactor / copy paste error.

willcl-ark commented on .github/silent-merge-test.py:
```diff
@@ -0,0 +1,166 @@
+    start_time = time.monotonic()
+
+    print("Fetching open PRs...")
+    open_prs = get_open_prs(repo)
```
Even though these have a `mergable` field, I think it's bunk :( :

```fish
core/worktrees/pr-33145 on  pr-33145 [$] via △ v4.1.2 via 🐍 v3.13.12 via ❄️  impure (nix-shell-env) took 57s
❯ gh api 'repos/bitcoin/bitcoin/pulls/35342' --jq '{number, mergeable, mergeable_state, keys: (keys | map(select(test("merge"))))}'
{
  "keys": [
    "auto_merge",
    "merge_commit_sha",
    "mergeable",
    "mergeable_state",
    "merged",
    "merged_at",
    "merged_by"
  ],
  "mergeable": true,
  "mergeable_state": "blocked",
  "number": 35342
}

core/worktrees/pr-33145 on  pr-33145 [$] via △ v4.1.2 via 🐍 v3.13.12 via ❄️  impure (nix-shell-env)
❯ gh api 'repos/bitcoin/bitcoin/pulls?state=open&per_page=1' --jq '.[0] | {number, mergeable, mergeable_state, keys: (keys | map(select(test("merge"))))}'
{
  "keys": [
    "auto_merge",
    "merge_commit_sha",
    "merged_at"
  ],
  "mergeable": null,
  "mergeable_state": null,
  "number": 35342
}
```

perhaps we need to iterate each pr individually

maflcko replied:
Jup, you'll need to call something like `get_pull_mergeable` in https://github.com/maflcko/DrahtBot/blob/6084f507bc64af4bfd5b3c33ebf725a57c66320a/rerun_ci/src/main.rs#L86

m3dwards replied:
Thanks for the heads up, that's a good catch. Will just check each PR individually, won't add much extra time.

willcl-ark reviewed (commented):
I feel like I've forgotten context here (sorry), but why can this job not just do something like:

- check out bitcoin/bitcoin:master
- fetch PR head
- attempt the merge locally
- run a fixed build command, something like:

```bash
rm -Rf build
cmake -B build --parallel
cmake --build build

# optionally
ctest --test-dir build
```

- publish the result?

this could probably be done on a bare runner host too, negating need for docker et al. I think it would probably be faster as well...

maflcko commented:
> I feel like I've forgotten context here (sorry), but why can this job not just do something like:
> 
>     * check out bitcoin/bitcoin:master
> 
>     * fetch PR head
> 
>     * attempt the merge locally
> 
>     * run a fixed build command, something like:
> 
> 
> ```shell
> rm -Rf build
> cmake -B build --parallel
> cmake --build build
> 
> # optionally
> ctest --test-dir build
> ```
> 
>     * publish the result?
> 
> 
> this could probably be done on a bare runner host too, negating need for docker et al. I think it would probably be faster as well...

Yeah, if that is less code, that works, too. For reference, except for `rm -rf`, you can just call `python3 .github/ci-test-each-commit-exec.py` to do all of the cmake stuff.

m3dwards commented:
> I feel like I've forgotten context here (sorry), but why can this job not just do something like:
> 
> * check out bitcoin/bitcoin:master
> * fetch PR head
> * attempt the merge locally
> * run a fixed build command, something like:
> 
> ```shell
> rm -Rf build
> cmake -B build --parallel
> cmake --build build
> 
> # optionally
> ctest --test-dir build
> ```
> 
> * publish the result?
> 
> this could probably be done on a bare runner host too, negating need for docker et al. I think it would probably be faster as well...

I'm open to changing it to this but this is more code than it currently is which is simply calling `./ci/test_run_all.sh`. I also quite like the idea that it's currently running the functional tests because a failure could be logical rather than just a compilation error.

maflcko commented:
> functional tests

`.github/ci-test-each-commit-exec.py` does run them, see the prior comment.

m3dwards commented:
> > functional tests
> 
> `.github/ci-test-each-commit-exec.py` does run them, see the prior comment.

I'm not really understanding the advantages of this over `./ci/test_run_all.sh`. We don't want to check each commit with this job, only the merge commit right?

maflcko commented:
The benefit would be that it doesn't require docker, so I think the three `uses: ./.github/actions` can be dropped. `.github/ci-test-each-commit-exec.py` only tests a single commit, the name can be changed, if needed.

But either way is fine. It should be trivial to adjust later, if needed.

maflcko commented:
> More concerned with the cache actions changing but can deal with that if and when it happens.

This was fixed in https://github.com/bitcoin/bitcoin/pull/35348

m3dwards commented:
Pushed latest changes to show direction. Running a bunch of tests on my fork. Will switch this PR to ready to review when the local testing is complete. It's just a bit time consuming as you have to do multiple full CI runs on a bunch of fake PRs.

maflcko commented on .github/silent-merge-test.py:
```diff
@@ -0,0 +1,217 @@
+            print(f"PR #{pr['number']} has merge conflicts against current master, skipping.")
+            continue
+
+        ci_result = run([".github/ci-test-each-commit-exec.py"], check=False,
```
Just checked my rerun_ci.sh script, and realized I was also running `--task 'lint'`. lint should be fast, so it could also be run here?

m3dwards replied:
Do you have a link to that script? To run lint as it is in `ci.yml` would require pulling docker in right and all the docker caches etc etc.

maflcko replied:
Well the script is  https://github.com/maflcko/DrahtBot/blob/2038d4d541b89bf2601614b5656d7d30d73e17be/rerun_ci/src/main.rs#L15

I think you can just call `ci/lint.py`, because docker is already installed and docker has a built-in cache that works out of the box without having to pull anything?

m3dwards replied:
I've run experiments today adding lint and it dramatically slowed down how long per PR it takes to process. Right now it's shockingly fast at about 5-7 minutes per PR with a warm ccache. Adding lint took this to 35 minutes per PR. I did expect the first one to be slower but expected the second PR checked to be faster as the image would already have been built?

I think I'm of the mind to just leave lint out as I wouldn't expect a silent merge issue to be often caught by the lint checker and it just slows things down and adds more complexity.

That said, if you are keen to have it, I can keep working to try and figure out why it's so slow and if we can do better caching etc.

Just for reference here is a 2 PR run that was done with lint (1 hour 16 minutes): https://github.com/m3dwards/bitcoin/actions/runs/27784000378

And one done without lint (23 minutes): https://github.com/m3dwards/bitcoin/actions/runs/27787578648

maflcko replied:
Ok, fair to leave it for later.

Also, looking at the logs, it could be a bit tedious to find the failure, as it is hidden in the middle of a long log? Also, there are no timestamps on any of the lines at all?

Probably fine, but maybe the CDash instance can be used to submit each result, so that each result is stand-alone?

m3dwards replied:
I can add timestamps.

The logs are only printed when a PR fails but if multiple PRs fail which I would think would be unlikely then yes you would have to differentiate between the PRs and search for your PR number in the log output. I don't think this is very likely though as silent merge issues should be rare-ish?

Is that good enough for this PR as it's probably mostly the same as existing CI.

I am happy to look at adding results to CDash unless you think with only printing results when a PR fails is sufficient, or that it could be added as a followup.

maflcko reviewed (commented):
This is still in Draft. What is the status here?

m3dwards commented:
> This is still in Draft. What is the status here?

I think pretty much ready for review. I'm still running some tests in the background and I can add the lint job you mentioned.

maflcko reviewed (commented):
lgtm. Seems fine to put the code in this repo and call it inside this CI. An alternative would be to have the code under DrahtBot and run it there.

m3dwards commented:
Added timestamps to the log messages if they do not already have them (such as some of the functional test messages). Timestamps were also only added to the failure log as all other log messages are timestamped properly by GHA.

The logs are now collapsed into their own section meaning it should be much easier to find the log messages when something is wrong. For this reason and because the functional tests aren't integrated with ctest at the moment I have left out CDash integration either for a follow up or for never.

@willcl-ark prototyped CDash here https://github.com/bitcoin/bitcoin/compare/master...willcl-ark:bitcoin:pr-33145 but because of the above I have chosen not to add this to this PR.

Example of the logs being collapsed:

<img width="465" height="100" alt="Screenshot 2026-07-01 at 16 26 25" src="https://github.com/user-attachments/assets/b20d6f61-9f14-42a1-a79e-96ab2f38cdee" />

sedited commented:
This hasn't received attention in a while. Can you take another look @willcl-ark @maflcko?
